lib/elixir/src/elixir_erl.erl

```diff
@@ -375,10 +375,21 @@ specs_form(Map, Data, Defmacro, Unreachable, Forms) ->
   Optional = lists:flatten(take_type_spec(Data, optional_callbacks)),
   SpecsForms = specs_form(spec, Specs, Unreachable, [], Defmacro, Forms),
   AllCallbacks = Callbacks ++ Macrocallbacks,
+  validate_behaviour_info_and_attributes(Map, AllCallbacks),
   validate_optional_callbacks(Map, AllCallbacks, Optional),
   specs_form(callback, AllCallbacks, [], Optional,
              [NameArity || {_, NameArity, _, _} <- Macrocallbacks], SpecsForms).
 
+validate_behaviour_info_and_attributes(#{definitions := Defs, module := Mod} = Map, AllCallbacks) ->
+  case {lists:keyfind({behaviour_info, 1}, 1, Defs), AllCallbacks} of
+    {false, _} ->
+      ok;
+    {_, [{Kind, {Name, Arity}, _, _} | _]} when Kind == callback; Kind == macrocallback ->
+      form_error(Map, {behaviour_info, {Mod, Name, Arity}});
```

Comment:
We don't need to keep the same message shape. :)

Reply:
Oh cool, okay. I don't know why I thought we did. I'll change it then!

Reply:
Also, do we need to make sure to do nothing when we receive this message in `elixir_errors.erl`?

Reply:
@josevalim ah yes, probably. I saw that adding the error here made this error pop up and didn't think about doing nothing for the old error.

Reply:
@josevalim since ours pops up before, I think this PR is ready to go then!